### PR TITLE
NXBT-3607: Add Nuxeo build and Web UI release validation pipeline

### PR DESCRIPTION
**Pull request by tfowley**

### CODE REVIEW

File: nuxeo-server-web-ui/Docker/skaffold.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#
+apiVersion: skaffold/v2beta17
+kind: Config
+build:
+  tagPolicy:
+    envTemplate:
+      template: "{{.VERSION}}"
+  artifacts:
+    - image: "$DOCKER_REGISTRY/nuxeo/server-build-web-ui-release"
+      context: server-build-web-ui-release-docker
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Docker/skaffold.yaml]: this is the parent path of the Dockerfile from the repository root

File: nuxeo-server-web-ui/Docker/skaffold.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        dockerfile: Dockerfile
+        buildContext:
+          localDir: { }
+        buildArgs:
+          BASE_IMAGE: "{{.PRIVATE_DOCKER_REGISTRY}}/nuxeo/nuxeo:{{.NUXEO_IMAGE_VERSION}}"
+          BUILD_TAG: "{{.BUILD_TAG}}"
+          SCM_REF: "{{.SCM_REF}}"
+          VERSION: "{{.VERSION}}"
+          CLID: "{{.CLID}}"
+          PACKAGES: "nuxeo-web-ui:3.0.10"
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Docker/skaffold.yaml]: add TODO: will be NUXEO_WEB_UI_VERSION

File: nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+          Build Docker image
+          ------------------------------------------
+          Image tag: ${VERSION}
+          Registry: ${DOCKER_REGISTRY}
+          """
+          withEnv(["NUXEO_IMAGE_VERSION=${getNuxeoImageVersion()}"]) {
+            withCredentials([string(credentialsId: 'instance-clid', variable: 'INSTANCE_CLID')]) {
+              script {
+                // build and push Docker image to the Jenkins X internal Docker registry
+                env.DOCKER_PATH = 'server-build-web-ui-release'
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy]: use a Groovy var: def dockerPath = 'docker'

File: nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                sh '''#!/bin/bash +x
+                  CLID=\$(echo -e "${INSTANCE_CLID}" | sed ':a;N;\$!ba;s/\\n/--/g') skaffold build -f $DOCKER_PATH/skaffold.yaml~gen
+                '''
+                def image = "${DOCKER_REGISTRY}/nuxeo/${DOCKER_IMAGE_NAME}:${VERSION}"
+                // using test in skaffold.yaml doesn't seem to work with a remote image,
+                // despite https://github.com/GoogleContainerTools/skaffold/issues/3907 supposedly solved
+                sh """
+                  docker pull ${image}
+                  container-structure-test test --image ${image} --config ${DOCKER_PATH}/test/*
+                """
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy]: unneeded

File: nuxeo-server-web-ui/Docker/skaffold.yaml [hunk 5]
```diff
@@ -0,0 +1,48 @@
+# (C) Copyright 2019-2020 Nuxeo (http://nuxeo.com/) and others.
+#
+# Contributors:
+#     Kevin Leturc <kleturc@nuxeo.com>
+#     Anahide Tchertchian
+#     Arnaud Kervern
+#
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Docker/skaffold.yaml]: Either remove or fix: Copyright 2022, you're name as contributor.

File: nuxeo-server-web-ui/Docker/skaffold.yaml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+apiVersion: skaffold/v2beta17
+kind: Config
+build:
+  tagPolicy:
+    envTemplate:
+      template: "{{.VERSION}}"
+  artifacts:
+    - image: "$DOCKER_REGISTRY/nuxeo/server-build-web-ui-release"
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Docker/skaffold.yaml]: I think we can have a generic image name and use the tag for the server and web-ui versions.
Something like:
- xxx/nuxeo/server-web-ui:2021.18.3-3.0.10
- xxx/nuxeo/server-web-ui:2021.17-3.0.11-rc01
- xxx/nuxeo/server-web-ui:2021.18.3-3.0.11-rc01

File: nuxeo-server-web-ui/Docker/skaffold.yaml [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+      context: server-build-web-ui-release-docker
+      kaniko:
+        useNewRun: true
+        singleSnapshot: true
+        snapshotMode: "time"
+        dockerfile: Dockerfile
+        buildContext:
+          localDir: { }
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Docker/skaffold.yaml]: unneeded

File: nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+*     Thomas Fowley
+*/
+
+String getMavenArgs() {
+  def args = '-B -nsu -Dnuxeo.skip.enforcer=true -P-nexus,nexus-private'
+  if (!isPullRequest()) {
+    args += ' -Prelease'
+  }
+  return args
+}
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy]: won't be needed

File: nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+void getNuxeoImageVersion() {
+  return sh(returnStdout: true, script: 'mvn org.apache.maven.plugins:maven-help-plugin:3.1.0:evaluate -Dexpression=nuxeo.platform.version -q -DforceStdout').trim();
+}
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy]: we will probably get the Nuxeo image version from a job parameter, see some other jobs as example

File: nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+pipeline {
+  agent {
+    label 'jenkins-nuxeo-web-ui'
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy]: nuxeo-web-ui-ftests, see https://github.com/nuxeo/platform-ci/pull/49#discussion_r829549153

File: nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+  }
+  options {
+    buildDiscarder(logRotator(daysToKeepStr: '60', numToKeepStr: '60', artifactNumToKeepStr: '5'))
+  }
+  environment {
+    DEFAULT_CONTAINER = 'ftests'
+    MAVEN_ARGS = getMavenArgs()
+    DOCKER_IMAGE_NAME = 'server-build-web-ui-release'
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy]: server-web-ui

File: nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+pipeline {
+  agent {
+    label 'jenkins-nuxeo-web-ui'
+  }
+  options {
+    buildDiscarder(logRotator(daysToKeepStr: '60', numToKeepStr: '60', artifactNumToKeepStr: '5'))
+  }
+  environment {
+    DEFAULT_CONTAINER = 'ftests'
+    MAVEN_ARGS = getMavenArgs()
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy]: unneeded

File: nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    DOCKER_IMAGE_NAME = 'server-build-web-ui-release'
+    SLACK_CHANNEL = 'platform-notifs'
+    CONNECT_PREPROD_URL = 'https://nos-preprod-connect.nuxeocloud.com/nuxeo'
```
Comment (ataillefer) [inline nuxeo-server-web-ui/Jenkinsfiles/server-build-web-ui-release.groovy]: unneeded

**Review by ataillefer — COMMENTED** [review]
Please rename Docker to docker.

**Comment by ataillefer** [comment]
As mentioned in https://github.com/nuxeo/platform-ci/pull/49#discussion_r829548567, let's use nuxeo-web-ui as a root folder, to match the Nuxeo package name.

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+* Contributors:
+*     Thomas Fowley
+*/
+CONNECT_PROD_URL = 'https://connect.nuxeo.com/nuxeo/site/'
+CONNECT_PREPROD_URL = 'https://nos-preprod-connect.nuxeocloud.com/nuxeo/site/'
```
Comment (kevinleturc) [inline nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy]: These variables should be available in the environment, see https://github.com/nuxeo/platform-ci/blob/master/charts/jenkins/values-casc-jenkins.yaml.gotmpl#L19-L22

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+WEB_UI_BRANCH = 'maintenance-3.0.x'
+TEST_ENVIRONMENT = 'default'
+TEST_HELM_RELEASE = 'nuxeo'
+
+properties([
+  [$class: 'GithubProjectProperty', projectUrlStr: 'https://github.com/nuxeo/nuxeo-integration-tests'],
+  [$class: 'BuildDiscarderProperty', strategy: [$class: 'LogRotator', daysToKeepStr: '60', numToKeepStr: '60', artifactNumToKeepStr: '5']],
+  disableConcurrentBuilds(),
+])
+def cloneRepo(name, branch, relativePath = name) {
```
Comment (kevinleturc) [inline nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy]:
```suggestion
])

def cloneRepo(name, branch, relativePath = name) {
```
And between all other methods.

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+def getConnectURL(isBuild) {
+  return isBuild ? CONNECT_PREPROD_URL : CONNECT_PROD_URL
+}
+
+def getWebUIPackageVersion(packageName, isBuild, nuxeoVersion) {
+  def targetPlatform = "lts-${nuxeoVersion}"
+  // e.g. "nuxeo-web-ui:3.0.8-rc.001" (preprod) or "nuxeo-web-ui:3.0.7" (prod)
+  def credentialsId = isBuild ? 'connect-preprod' : 'connect-prod'
+  withCredentials([usernameColonPassword(credentialsId: credentialsId, variable: 'CONNECT_AUTH')]) {
+    def connectURL = isBuild ? CONNECT_PREPROD_URL : CONNECT_PROD_URL
```
Comment (kevinleturc) [inline nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy]:
```suggestion
    def connectURL = getConnectURL()
```

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+     steps {
+       setGitHubBuildStatus('docker/build', 'Build Docker image', 'PENDING')
+       container('ftests') {
+         echo """
+         ------------------------------------------
+         Build nuxeo-web-ui-ftest-sample package
+         ------------------------------------------
+         """
+         echo 'Build package'
+         dir(NUXEO_WEB_UI_CLONE){
```
Comment (kevinleturc) [inline nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy]:
```suggestion
         dir(NUXEO_WEB_UI_CLONE) {
```

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+                """
+              }
+              setGitHubBuildStatus('docker/deploy', 'Deploy docker image', 'SUCCESS')
+            } catch (e) {
+              archiveScreenshots()
+              setGitHubBuildStatus('docker/deploy', 'Deploy docker image', 'FAILURE')
+              throw e
+            } finally {
+              try {
+                cucumber buildStatus: 'FAILURE',
```
Comment (kevinleturc) [inline nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy:280]: Shouldn't the buildStatus depend on the actual result of `npm run ftest`?

File: nuxeo-web-ui/docker/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,30 @@
+# vim:set ft=dockerfile:
+ARG BASE_IMAGE
+
+FROM ${BASE_IMAGE}
+
+ARG VERSION
+ARG SCM_REF
+ARG BUILD_TAG
+ARG CLID
+ARG PACKAGES
```
Comment (kevinleturc) [inline nuxeo-web-ui/docker/Dockerfile]: It doesn't seem needed.

File: nuxeo-web-ui/docker/skaffold.yaml [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+# Contributors:
+#     Thomas Fowley
+apiVersion: skaffold/v2beta17
+kind: Config
+build:
+  tagPolicy:
+    envTemplate:
+      template: "{{.TEST_IMAGE_VERSION}}"
+  artifacts:
+    - image: "$DOCKER_REGISTRY/nuxeo/nuxeo-web-ui"
```
Comment (kevinleturc) [inline nuxeo-web-ui/docker/skaffold.yaml]: I would put ftests somewhere as this image contains a ftests marketplace, wdyt?

File: nuxeo-web-ui/helm/values/values-elasticsearch.yaml.gotmpl [hunk 21]
```diff
@@ -0,0 +1,30 @@
+image: "docker.elastic.co/elasticsearch/elasticsearch-oss"
+clusterName: elasticsearch
+replicas: 1
+minimumMasterNodes: 1
+resources:
+  requests:
+    cpu: "250m"
+    memory: "1024Mi"
```
Comment (kevinleturc) [inline nuxeo-web-ui/helm/values/values-elasticsearch.yaml.gotmpl]: Maybe ?
```suggestion
    memory: "1G"
```

File: nuxeo-web-ui/helm/values/values-elasticsearch.yaml.gotmpl [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+  limits:
+    cpu: "500m"
+    memory: "1024Mi"
```
Comment (kevinleturc) [inline nuxeo-web-ui/helm/values/values-elasticsearch.yaml.gotmpl]: Maybe ?
```suggestion
    memory: "1G"
```

**Review by kevinleturc — COMMENTED** [review]
Great work 👍

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy (hunk 18 above)
Comment (tfowley) [reply nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy:280]: Yes, the buildStatus depends on the report which in turn is set by the results of `npm run ftest`. 
`Type: String
buildStatus (optional)
Build [result](http://javadoc.jenkins-ci.org/hudson/model/Result.html) to which the build should be set when the report is marked as failed or unstable.`
(from https://www.jenkins.io/doc/pipeline/steps/cucumber-reports/)

**Review by kevinleturc — DISMISSED** [review]

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy [hunk 23]
```diff
@@ -0,0 +1,328 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+* Contributors:
+*     Thomas Fowley
+*/
+NUXEO_MAJOR_VERSION = '2021'
+DEFAULT_CONTAINER = 'ftests'
```
Comment (ataillefer) [inline nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy:20]: unused, either use it in the `container()` steps or remove it

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy [hunk 24]
```diff
@@ -0,0 +1,328 @@
...
+        }
+      }
+    }
+  }
+  post {
+    success {
+      script {
+        if (env.DRY_RUN != 'true'
+          && !hudson.model.Result.SUCCESS.toString().equals(currentBuild.getPreviousBuild()?.getResult())){
+          slackSend(channel: "${SLACK_CHANNEL}", color: 'good', message: "Nuxeo-Integration-Tests: All tests passed for Nuxeo:${NUXEO_IMAGE_VERSION} & Web UI:${NUXEO_WEB_UI_VERSION}. #${BUILD_NUMBER}: ${BUILD_URL}")
```
Comment (ataillefer) [inline nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy]:
```suggestion
          slackSend(channel: "${SLACK_CHANNEL}", color: 'good', message: "Nuxeo-Integration-Tests: All tests passed for Nuxeo: ${NUXEO_IMAGE_VERSION} & Web UI: ${NUXEO_WEB_UI_VERSION}. #${BUILD_NUMBER}: ${BUILD_URL}")
```

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy [hunk 25]
```diff
@@ -0,0 +1,328 @@
...
+        }
+      }
+    }
+    unsuccessful {
+      script {
+        if (env.DRY_RUN != 'true'
+          && !hudson.model.Result.SUCCESS.toString().equals(currentBuild.getPreviousBuild()?.getResult())){
+          slackSend(channel: "${SLACK_CHANNEL}", color: 'danger', message: "Nuxeo-Integration-Tests: Tests failing for Nuxeo:${NUXEO_IMAGE_VERSION} & Web UI:${NUXEO_WEB_UI_VERSION}. #${BUILD_NUMBER}: ${BUILD_URL}")
```
Comment (ataillefer) [inline nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy]:
```suggestion
          slackSend(channel: "${SLACK_CHANNEL}", color: 'danger', message: "Nuxeo-Integration-Tests: Tests failing for Nuxeo: ${NUXEO_IMAGE_VERSION} & Web UI: ${NUXEO_WEB_UI_VERSION}. #${BUILD_NUMBER}: ${BUILD_URL}")
```

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy [hunk 26]
```diff
@@ -0,0 +1,328 @@
...
+            doGenerateSubmoduleConfigurations: false,
+            extensions: [
+              [$class: 'RelativeTargetDirectory', relativeTargetDir: relativePath],
+              [$class: 'WipeWorkspace'],
+              [$class: 'CloneOption', depth: 0, noTags: false, reference: '', shallow: false, timeout: 60],
+              [$class: 'CheckoutOption', timeout: 60],
+              [$class: 'LocalBranch']
+            ],
+            submoduleCfg: [],
+            userRemoteConfigs: [[credentialsId: 'jx-pipeline-git-github-git', url: 'https://github.com/nuxeo/' + name]]
```
Comment (ataillefer) [inline nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy]: can you please fix the indentation

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy [hunk 27]
```diff
@@ -0,0 +1,328 @@
...
+         Build Docker image
+         ------------------------------------------
+         Image tag: ${TEST_IMAGE_VERSION}
+         Registry: ${DOCKER_REGISTRY}
+         Nuxeo image version: ${NUXEO_IMAGE_VERSION}
+         Web UI package version: ${NUXEO_WEB_UI_VERSION}
+         """
+         withCredentials([string(credentialsId: 'instance-clid', variable: 'INSTANCE_CLID')]) {
+           script {
+             // build and push Docker image to the Jenkins X internal Docker registry
```
Comment (ataillefer) [inline nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy]: can be removed

File: nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy [hunk 28]
```diff
@@ -0,0 +1,328 @@
...
+              }
+              setGitHubBuildStatus('docker/deploy', 'Deploy docker image', 'SUCCESS')
+            } catch (e) {
+              archiveScreenshots()
+              setGitHubBuildStatus('docker/deploy', 'Deploy docker image', 'FAILURE')
+              throw e
+            } finally {
+              try {
+                cucumber buildStatus: 'FAILURE',
+                  reportTitle: "Test report WebUI:${NUXEO_WEB_UI_VERSION} Nuxeo:${NUXEO_IMAGE_VERSION}",
```
Comment (ataillefer) [inline nuxeo-web-ui/Jenkinsfiles/server-web-ui.groovy]:
```suggestion
                  reportTitle: "Ftests - Nuxeo: ${NUXEO_IMAGE_VERSION}, Web UI: ${NUXEO_WEB_UI_VERSION}",
```

**Review by ataillefer — COMMENTED** [review]
:muscle:

**Review by ataillefer — APPROVED** [review]

**Review by kevinleturc — APPROVED** [review]

**Pull request merged**